javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,230 @@
+  let hashCode = 0;
+
+  const primeNumber = 31;
+  for (let i = 0; i < string.length; i++) {
+    hashCode = primeNumber * hashCode + string.charCodeAt(i);
+  }
+
+  return hashCode;
+}
+```
+
+With our new function we will have different hash codes for the names `"Sara"` and `"raSa"` that is because even if both names have the same letters, some of the letters appear in different locations. The hash code started to change because we are multiplying the old hash every new iteration and then adding the letter code.
+
+<div class="lesson-note lesson-note--tip" markdown="1">
+  Notice the usage of prime number. We could have chosen any number we wanted, but prime numbers are even better to introduce less hash codes that are divisible by the same bucket length, which will make collisions less likely to happen.
+</div>
+
+Even though we reworked our hash function to avoid the `"Sara"/"raSa"` collision, there is always the possibility for collisions. Because we have a finite number of buckets, there is no way to eliminate collisions entirely. But we can try to minimize them as much as possible.
+
+#### Dealing with collisions
+
+Enters `Linked Lists`. If each `Node` inside the bucket is also a Linked List, Then we look for bucket `508` if it's empty we insert the head of Linked List, If a Node head exists in a bucket we follow that Linked List to add to the end of it.
+
+You probably understand by this point why we must write a good hashing function which eliminates as many collisions as possible. Most likely you will not be writing your own hash functions, as most languages have it built in. But understanding how hash functions work is important.
+
+Check out [This Video](https://www.youtube.com/watch?v=btT4bCOvqjs) from CS50 that explains the concept shortly with visualization
+
+### Growth of a hash table
+
+Let's talk about the growth of our buckets. We don't have infinite memory, so we can't have an infinite amount of buckets. We need to start somewhere, but starting too big is also a waste of memory if we're only going to have a hash map with a single value in it. So to deal with this issue, we should start with a small array for our buckets. We'll use an array with 10 elements and indexes from 0 to 9.
+
+<div class="lesson-note lesson-note--tip" markdown="1">
+  Most programming languages start with the default size of `16` because it's a power of 2, which help with some techniques for performance that require bit manipulation for indexes. But for this example, we will be using a starting size of 10.
+</div>
+
+How are we going to insert into those buckets when our hash function generates big numbers like `20353924`? We make use of the modulo `%` operation `given any number modulo by 10 we will get a number between 0 and 9`.
+For example, if we are to find the bucket where the value `"Manon"` will land, then we do the following:
+![hashing using hash code and modular operation example](./project_hash_map/imgs/01.png)
```

Comment:
This paragraph doesn't contribute to the "Growth of hash tables" section. Maybe move it to "Buckets" section?

Reply:
The Buckets section does not introduce the initial size of our array, I had to add this to clear a confusion of how are we going to translate big numbers when our array is only limited to 16? If I need to move this to the buckets section, then I will have to find a way to introduce the array size there as well.

Any suggestions?

Reply:
Nice observation. Then, I propose the following wording suggestion:
```diff
- How are we going to insert into those buckets when our hash function generates big numbers like `20353924`? We make use of the modulo `%` operation `given any number modulo by 10 we will get a number between 0 and 9`.
+ Now that we have imposed the restriction of finite buckets, how will we deal with hash codes that exceed the available indexes in our hash map like `20353924`.  We make use of the modulo `%` operator. 
```